game/src/common/waypoints.rs

```diff
@@ -209,76 +192,45 @@ impl InputWaypoints {
         false
     }
 
-    pub fn draw(&self, g: &mut GfxCtx) {
-        g.redraw(&self.draw_waypoints);
-        g.redraw(&self.draw_hover);
-    }
-
     fn get_waypoint_color(&self, idx: usize) -> Color {
         let total_waypoints = self.waypoints.len();
         match idx {
             0 => Color::GREEN,
             idx if idx == total_waypoints - 1 => Color::RED,
-            // technically this includes the case where idx >= total_waypoints which should hopefully never happen
             _ => [Color::BLUE, Color::ORANGE, Color::PURPLE][idx % 3],
         }
     }
 
-    fn update_waypoints_drawable(&mut self, ctx: &mut EventCtx) {
-        let mut batch = GeomBatch::new();
-        for (idx, waypt) in self.waypoints.iter().enumerate() {
+    /// The caller is responsible for calling `initialize_hover` and `rebuilt_during_drag`.
+    pub fn rebuild_world<T: ObjectID, F: Fn(WaypointID) -> T>(
+        &self,
+        ctx: &mut EventCtx,
+        world: &mut World<T>,
+        wrap_id: F,
+        zorder: usize,
+    ) {
+        for (idx, waypoint) in self.waypoints.iter().enumerate() {
+            let hitbox = Circle::new(waypoint.center, Distance::meters(30.0)).to_polygon();
             let color = self.get_waypoint_color(idx);
-            let text = get_waypoint_text(idx);
 
-            let mut geom = GeomBatch::new();
-            geom.push(color, waypt.hitbox.clone());
-            geom.append(
-                Text::from(Line(format!("{}", text)).fg(Color::WHITE))
+            let mut draw_normal = GeomBatch::new();
+            draw_normal.push(color, hitbox.clone());
+            draw_normal.append(
+                Text::from(Line(get_waypoint_text(idx).to_string()).fg(Color::WHITE))
                     .render(ctx)
-                    .centered_on(waypt.hitbox.center()),
+                    .centered_on(waypoint.center),
             );
 
-            batch.append(geom);
-        }
-        self.draw_waypoints = ctx.upload(batch);
-    }
-
-    fn update_hover(&mut self, ctx: &EventCtx) {
-        self.hovering_on_waypt = None;
-
-        if let Some(pt) = ctx.canvas.get_cursor_in_map_space() {
-            self.hovering_on_waypt = self
-                .waypoints
-                .iter()
-                .position(|waypt| waypt.hitbox.contains_pt(pt));
-        }
-
-        let mut batch = GeomBatch::new();
-        if let Some(idx) = self.hovering_on_waypt {
-            batch.push(Color::BLUE.alpha(0.5), self.waypoints[idx].hitbox.clone());
+            world
```

Comment:
Just compare the code before/after here. The declarative builder API is so much easier to understand!